pubsub/iterator.go

```diff
@@ -309,6 +329,24 @@ func (it *messageIterator) receive(maxToPull int32) ([]*Message, error) {
 				pendingMessages[ackID] = m
 			}
 		}
+
+		if it.enableTracing {
+			ctx := context.Background()
+			if m.Attributes != nil {
+				ctx = propagation.TraceContext{}.Extract(ctx, newMessageCarrier(m))
+			}
+			attr := getSubscriberOpts(it.projectID, it.subID, m)
+			_, span := startSpan(ctx, subscribeSpanName, it.subID, attr...)
+			span.SetAttributes(
+				attribute.Bool(eosAttribute, it.enableExactlyOnceDelivery),
+				attribute.String(ackIDAttribute, ackID),
+				semconv.MessagingBatchMessageCount(len(msgs)),
+				semconv.CodeFunction("receive"),
+			)
+			if span.SpanContext().IsSampled() {
```

Comment:
I've been trying to diagnose broken traces and I believe this is the problem. Let's say we have a trace that's `HTTP -> RPC -> Pub/Sub Publish -> Pub/Sub Receive` and it's not sampled because in `HTTP` we sample at 0.1% which means when it gets to here it won't be stored in `activeSpans`.

Later in `*Subscription.Receive` it does `c, ok := iter.activeSpans.Load(ackh.ackID)` and then re-creates a context with the originally-received span but since it wasn't in `activeSpans` it will end up with a span-less context. This means when our handler gets the context and creates a new span it has no context and hits our sampler which might choose to sample it.

This results in spans that seem to start with our handler out of nowhere rather than connecting back to the original `HTTP` span.

Reply:
> This means when our handler gets the context and creates a new span it has no context and hits our sampler which might choose to sample it.

Hm, this is a good point. I originally envisioned it wasn't necessary to store the span contexts of non-sampled spans, since if you use Parent based sampling, none of these spans would be useful. I forgot this is necessary to propagate the `IsSampled == false` status to the handler. Let me remove this check and always store the span contexts even for `IsSampled == false` spans.